File: bindings/libotp-php/src/PHPOTP/Otp/Rfc4226.php
<?php
/**
 * Created by JetBrains PhpStorm.
 * User: luissantos
 * Date: 1/26/14
 * Time: 7:40 PM
 * To change this template use File | Settings | File Templates.
 */

namespace PHPOTP\Otp;


use PHPOTP\Authenticator\Authenticator;

class Rfc4226 {

    /**
     * @var Authenticator
     */
    protected $authenticator;

    /**
     * @var int
     */
    protected $digit;


    function __construct(Authenticator $authenticator,$digit = 6)
    {
        $this->authenticator = $authenticator;
        $this->digit = $digit;
    }


    protected function StToNum($binary_data){
        return hexdec(bin2hex($binary_data));
    }

    /**
     * @return \PHPOTP\Authenticator\Authenticator
     */
    protected  function getAuthenticator()
    {
        return $this->authenticator;
    }


    public  function generateOtp($counter){

        $hmac = $this->hmac($this->normalizeNumber($counter));

        $SBits = $this->dynamicTruncation($hmac);

        $Snum = $this->StToNum($SBits);


        return $Snum % (pow(10,$this->digit));
    }

    protected function hmac($counter){

        $auth = $this->getAuthenticator();

        return $auth->hash($counter);

    }


    protected function getOffsetBits($string){

        $last_byte = substr($string,-1);

        $last_byte_int = $this->bin2dec($last_byte);


        return $this->dec2bin($last_byte_int & 0x0f);
    }

    public function dynamicTruncation($string){

        $offsetbits = $this->getOffsetBits($string);

        $Offset = $this->StToNum($offsetbits);

        //Let P = String[OffSet]...String[OffSet+3]
        $p = substr($string,$Offset,4);


        /* Return the Last 31 bits of P */
        return $this->dec2bin($this->bin2dec($p) & bindec("01111111111111111111111111111111"));
    }

    /**
     * converts binary data to decimal
     *
     * @param $data
     *
     * @return number
     */
    protected  function  bin2dec($data){
        return hexdec(bin2hex($data));
    }

    /**
     *
     * converts decimal to binary data
     *
     * @param $number
     *
     * @return string
     */
    protected function dec2bin($number){
        $hex = dechex($number);
        $size = strlen($hex);

        return hex2bin(str_pad($hex,($size + ($size % 2)),"0",STR_PAD_LEFT));
    }

    /**
     *
     * Converts integer to binary and add padding zeros
     *
     * @param $number
     *
     * @return string
     */
    protected function normalizeNumber($number){
        return hex2bin(str_pad(dechex($number),"16","0",STR_PAD_LEFT));
    }

}

File: bindings/libotp-java/src/main/java/com/libotp/Otp/Rfc4226.java
package com.libotp.Otp;

import com.libotp.Authenticator.Authenticator;

import java.nio.ByteBuffer;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;

/**
 * Created by luissantos on 2/4/14.
 */
public class Rfc4226 {


    protected Authenticator authenticator;

    protected Integer digits;

    public Rfc4226(Authenticator authenticator, Integer digits) {
        this.authenticator = authenticator;
        this.digits = digits;
    }

    public Rfc4226(Authenticator authenticator) {
        this(authenticator,6);
    }

    public Authenticator getAuthenticator() {
        return authenticator;
    }

    public void setAuthenticator(Authenticator authenticator) {
        this.authenticator = authenticator;
    }

    public Integer getDigits() {
        return digits;
    }

    public void setDigits(Integer digits) {
        this.digits = digits;
    }


    protected Integer stToNum(byte[] data){
        return ByteBuffer.wrap(data).getInt();
    }

    public Integer generateOtp(Integer counter){

        return 0;
    }

    protected byte[] hmac(Integer counter) throws InvalidKeyException, NoSuchAlgorithmException {
        return authenticator.hash(counter);
    }

    protected byte[] getOffsetBits(byte[] data){

        byte offset = data[data.length-1];

        return new byte[]{ (byte)(offset & 0x0f) };

    }

    /*public  function generateOtp($counter){

        $hmac = $this->hmac($this->normalizeNumber($counter));

        $SBits = $this->dynamicTruncation($hmac);

        $Snum = $this->StToNum($SBits);


        return $Snum % (pow(10,$this->digit));
    }*/


    public byte[] dynamicTruncation(byte[] data){

        byte[] offsetbits = this.getOffsetBits(data);

        int Offset = this.stToNum(offsetbits);

        byte p[] = Arrays.copyOfRange(data,Offset,Offset+3);

        /* Return the Last 31 bits of P */
        return  this.intToByteArray(this.ByteArraytoInt(p) & 0x7FFFFFFF);
    }

    protected byte[] intToByteArray(int value){
        return ByteBuffer.allocate(4).putInt(value).array();
    }

    protected int ByteArraytoInt(byte[] data){
        return ByteBuffer.wrap(data).getInt();
    }

}
